src/components/gamespec/Board.js

```diff
@@ -5,18 +5,54 @@ import ItemTypes from './ItemTypes';
 import { Layer, Stage } from 'react-konva';
 
 import CanvasImage from './CanvasImage';
+import { piecesRef } from '../../firebase';
 
 const boxTarget = {
   drop(props, monitor, component) {
       let offset = monitor.getClientOffset(),
           item = monitor.getItem();
 
       let items = props.getItems();
+      let pieces = props.getPieces();
       let rect = component.refs.stage.getStage().getContainer().getBoundingClientRect();
       offset.x = offset.x - rect.left;
       offset.y = offset.y - rect.top;
+      let images = [];
+      items.forEach((element, index) =>{
```

Comment:
Keep single state variable, namely pieces which will look something like this:
```js
pieces = {
   images: [],
   offsets: [],
   key: ''
} 
```
Now you will fill them accordingly. So, in a piece image at ith index i.e. `piece.images[i]` will have offset at `piece,offsets[i]`.

The above shown information can be directly stored into firebase. But in an ideal case scenario you should only save images field in pieces for now.

After saving pieces in spec will look like:

```
pieces 
      key_here
               images: [ image_key_here, image_key2_here ]
```

Reply:
So you mean the pieces should belong to spec instead of being the same level with users, images and specs?

Reply:
Pieces is still a high level object in firebase, I have fixed the above description.

Reply:
one more question: should the image keys present in high level field--pieces be distinct or the keys can be duplicated?

Reply:
keys will the image keys present in the high level firebase field images. So, yes these image keys can come multiple times in single piece's images field.